### ISSUE TITLE
ApolloClient incorrectly parses the server response

### ISSUE THREAD

**Issue by jessedvrs** [apollostack/apollo-client#605]
TL;DR: There's a bug where apolloClient incorrectly parses the server response when you have an array with objects where multiple objects containing the same value for their `_id` field.

## Intended outcome:
I have the following server response from a GraphQL query. Note that `someArray` has two objects with the same `_id` and a different `order`.

```
{
    _id: "123",
    __typename: "SomeDocument",
    someArray: [
        {
            _id: "456",
            order: 1,
            __typename: "SomeSubDocument",
        },
        {
            _id: "456",
            order: 2,
            __typename: "SomeSubDocument",
        }
    ]
}
```

I expect `apolloClient.query()` to resolve exactly that.

## Actual outcome:
`apolloClient.query()` resolves the following object:

```
{
    _id: "123",
    someArray: [
        {
            _id: "456",
            order: 2
        },
        {
            _id: "456",
            order: 2
        }
    ]
}
```

Note that both array entries have `order: 2`. This only happens when the `_id` of the objects are the same.

## Reproduction:
- Your server should resolve an array with objects containing the same value for their `_id` field.
- It's important to name the field exactly `_id`. Otherwise the bug won't happen.
- The type of `_id` (defined server-side, in the schema) can be anything.

## Workaround:
For now, the workaround is to rename the field to `id` (server-side, in the schema).

**Comment by stubailo** [comment]
@jessedvrs are you using the Meteor integration? That adds an assumption that any object returned with the same `_id` will have the same fields.

In general, if you have a unique ID field in your API, I think it's a good idea to make sure the fields on that object are always the same, since any GraphQL cache, like Apollo and Relay, will assume that it is the same object.

**Comment by jessedvrs** [comment]
Yes I do, wow, it all makes sense now! Thanks a lot. From now on I'll only use '_id' in my schema where fields would logically have the same values.

**Comment by jessedvrs** [comment]
Still, the server resolves the correct result. So why would this Meteor integration perform any caching tricks *after* the server has responded with the full object as requested in the original query?

**Comment by stubailo** [comment]
The stuff you see in the network tab is just what the server returns. Before giving the result to the UI, Apollo Client processes it, puts it into a cache, then reads it out again. So the thing you get in the UI is actually what is in the cache, rather than the result of the network call directly.

**Comment by jessedvrs** [comment]
Yes that's what I suspected. Glad to see that one can configure this data object identifier in the apolloclient configuration. 'meteorClientConfig' seems to set the option 'dataIdFromObject: object.__typename + object._id' which does the trick, right?

**Comment by stubailo** [comment]
Yep, that's a Meteor-specific option picked in that package, which assumes you're using the `_id` from Mongo to identify objects. Works great for most cases!
